plugins/BEdita/API/src/Auth/EndpointAuthorize.php

```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * Get application for request.
+     *
+     * @return \BEdita\Core\Model\Entity\Application|null
+     * @throws \Cake\Network\Exception\ForbiddenException Throws an exception if API key is missing or invalid.
+     */
+    protected function getApplication()
+    {
+        $application = CurrentApplication::getApplication();
+        if ($application === null) {
+            try {
+                CurrentApplication::setFromApiKey(
+                    $this->request->getHeaderLine($this->_config['apiKeyHeaderName'])
+                );
+            } catch (\BadMethodCallException $e) {
+                throw new ForbiddenException(__d('bedita', 'Missing API key'));
+            } catch (RecordNotFoundException $e) {
+                throw new ForbiddenException(__d('bedita', 'Invalid API key'));
+            }
+
+            $application = CurrentApplication::getApplication();
+        }
+
+        return $application;
+    }
+
+    /**
+     * Get endpoint for request.
+     *
+     * @return \BEdita\Core\Model\Entity\Endpoint|null
+     */
+    protected function getEndpoint()
+    {
+        if (!empty($this->endpoint)) {
+            return $this->endpoint;
+        }
+
+        $endpointName = $this->request->url;
+        if (($slashPos = strpos($endpointName, '/')) !== false) {
+            $endpointName = substr($endpointName, 0, $slashPos);
+        }
+
+        if (empty($endpointName)) {
+            return null;
+        }
+
+        $this->endpoint = TableRegistry::get('Endpoints')->find()
+            ->where([
+                'Endpoints.name' => $endpointName,
+                'Endpoints.enabled' => true,
+            ])
+            ->first();
+
+        return $this->endpoint;
+    }
+
+    /**
+     * Get list of applicable permissions.
+     *
+     * @param mixed $user Authenticated (or anonymous) user.
+     * @param \BEdita\Core\Model\Entity\Application $application Current application.
+     * @param \BEdita\Core\Model\Entity\Endpoint|null $endpoint Current endpoint.
+     * @return \Cake\ORM\Query
+     * @todo Future optimization: Permissions that are `0` on the two bits that are interesting for the current request can be excluded...
+     */
+    protected function getPermissions($user, Application $application, Endpoint $endpoint = null)
+    {
+        $roleIds = Hash::extract($user, 'roles.{n}.id');
+        $applicationId = $application->id;
+        $endpointIds = $endpoint ? [$endpoint->id] : [];
+
+        return TableRegistry::get('EndpointPermissions')
+            ->find('byRole', compact('roleIds'))
+            ->find('byApplication', compact('applicationId'))
+            ->find('byEndpoint', compact('endpointIds'));
+    }
+
+    /**
+     * Checks if request can be authorized basing on a set of applicable permissions.
+     *
+     * @param \BEdita\Core\Model\Entity\EndpointPermission[] $permissions Set of applicable permissions.
+     * @return bool|string
```

Comment:
possible return values are true, false and `mine` - correct?
it's important to know that 'block' is never returned